subsys/net/lib/nrf_cloud/src/nrf_cloud_rest.c

```diff
@@ -131,29 +131,68 @@ static const char *const agps_req_type_strings[] = {
 	[NRF_CLOUD_REST_AGPS_REQ_CUSTOM]	= AGPS_REQ_TYPE_STR_CUSTOM,
 };
 
+/* Generate an authorization header value string in the form:
+ * "Authorization: Bearer JWT \r\n"
+ */
 static int generate_auth_header(const char *const tok, char **auth_hdr_out)
 {
-	if (!tok || !auth_hdr_out) {
+	if (!auth_hdr_out) {
+		LOG_ERR("Cannot generate auth header, no output pointer given.");
 		return -EINVAL;
 	}
 
-	int ret;
-	size_t buff_size = sizeof(AUTH_HDR_BEARER_TEMPLATE) + strlen(tok);
+	/* These lengths NOT including null terminators */
+	int tok_len = tok ? strlen(tok) : (CONFIG_MODEM_JWT_MAX_LEN - 1);
+	int prefix_len = sizeof(AUTH_HDR_BEARER_PREFIX) - 1;
+	int postfix_len = sizeof(CRLF) - 1;
 
-	*auth_hdr_out = k_malloc(buff_size);
+	*auth_hdr_out = k_malloc(prefix_len + tok_len + postfix_len + 1);
 	if (!*auth_hdr_out) {
 		return -ENOMEM;
 	}
-	ret = snprintk(*auth_hdr_out, buff_size, AUTH_HDR_BEARER_TEMPLATE, tok);
-	if (ret < 0 || ret >= buff_size) {
+
+	char *prefix_ptr = *auth_hdr_out;
+	char *tok_ptr = prefix_ptr + prefix_len;
+	char *postfix_ptr;
+	int ret = 0;
+
+	/* Write the prefix */
+	memcpy(prefix_ptr, AUTH_HDR_BEARER_PREFIX, prefix_len);
+
+	/* Copy the given token, or generate if none exists */
+	if (tok) {
+		memcpy(tok_ptr, tok, tok_len);
+	} else if (IS_ENABLED(CONFIG_NRF_CLOUD_REST_AUTOGEN_JWT)) {
```

Comment:
i believe `nrf_cloud_jwt_generate` (and `modem_jwt_generate`) will still get linked in if autogen is not enabled?

you may need to rewrite as:
```
	if (tok) {
		memcpy(tok_ptr, tok, tok_len);
#if defined(CONFIG_NRF_CLOUD_REST_AUTOGEN_JWT)
	} else {
		int err = nrf_cloud_jwt_generate(
			CONFIG_NRF_CLOUD_REST_AUTOGEN_JWT_LIFETIME,
			tok_ptr, tok_len + 1);

		if (err < 0) {
			LOG_ERR("Failed to auto-generate JWT, error: %d", err);
			ret = err;
			goto cleanup;
		}
		tok_len = strlen(tok_ptr);
#endif	
	}
```

Reply:
I am pretty sure that the block would be optimized away by the compiler before reaching the linking stage, if I understand correctly?

But the fact that I am not 100% confident in that answer, plus the fact that doing things this ways creates some complication with kconfig, probably means we should just switch the preprocessor defin approach anyways.

I'll do that.

Reply:
You can check the zephyr.map and zephyr.lst in build/zephyr to see whether it is linked or not.

Reply:
just checked and it does get optimized out 👯

Reply:
Doing a diffcheck on .map files with and without CONFIG_NRF_CLOUD_REST_AUTOGEN_JWT, I see a _lot_ of modem-jwt related references that are not present with CONFIG_NRF_CLOUD_REST_AUTOGEN_JWT=n

That would seem to confirm my understanding

That being stated, I'm still kind of on the fence about what the cleaner solution here is. Some of the other issues that have been brought up might make this issue moot

Reply:
yes, as long is this doesn't cause build errors for all valid configurations, it is fine.

Reply:
Thank you both for humoring me on this one.

As much as I prefer to use IS_ENABLED for the sake of code readability, the trickery required to avoid identifier declaration errors winds up harming readability more than IS_ENABLED aids it.

I will switch to just using plain preprocessor directives for this